docs: choosing the right storage for your use case (#808)
stellar/stellar-docs: docs/build/guides/archival/choosing-the-right-storage.mdx

---
title: How to choose the right storage type for your use case
hide_table_of_contents: true
description: This guide walks you through choosing the most suitable storage type for your use case and how to implement it.
---

State archival refers to how data on the network is stored. This data includes all information that comprises the network's current and historical state. Storage is managed by keeping the important data active and archiving the rest in order to help keep the network fast and cheap. Read more about state archival [here](../../../learn/encyclopedia/storage/state-archival.mdx).

State archival is vital for several reasons including:

> - It ensures that data is preserved efficiently and can be accessed when needed or restored in case of unexpected events;

Reviewer: ```suggestion
- It ensures that data is preserved efficiently and can be accessed when needed or restored if it becomes archived;
```

## Storage types

When a smart contract is deployed on the Stellar network, the uploaded Wasm bytecode is stored as a `CONTRACT_DATA` entry on the ledger. Each entry has a stipulated time it can be accessed on the ledger before it is either archived or permanently deleted from the network; this is known as the entry's time to live (TTL). The TTL is measured in ledgers. That is, if an entry's TTL is 50, the entry will be available for the next 50 ledgers. When the entry's TTL reaches zero, it is either deleted or archived based on the type of storage used.

There are three types of storage available on the Stellar network:

- Temporary storage
- Persistent storage
- Instance storage

### Temporary storage

As the name implies, temporary storage stores data for a short period. It is suitable for contracts that only require their data for a short while and can afford for the data to be discarded afterward without any consequences. Here, once the TTL reaches zero, the data will be deleted from the ledger and cannot be restored. Temporary storage has unlimited storage space and the cheapest fees among all the storage types.

> Temporary storage is best suited for contracts with easily replaceable data and contracts that are only relevant within a certain period. For example, a time-bound auction contract where users can place bids, which only need to be kept for the duration of the auction.

Reviewer: Another example or two of the type of data appropriate for temp storage would be really nice here! Not code, but just ideas to broaden the understanding of when it's appropriate.

    }

    // This function redeems points according to the user's balance
    pub fn redeem_points(env: Env, user: Address, points: u64) -> bool {
        let key = DataKey::Points(user.clone());
        let current_points: u64 = env.storage().persistent().get(&key).unwrap_or(0);

        if current_points >= points {
            let new_points = current_points - points;
            env.storage().persistent().set(&key, &new_points);
            true
        } else {
            false
        }
    }
}

```

### Instance storage

Instance storage is a type of persistent storage that is used for storing data associated with the contract instance itself; data that only needs to be accessed while the contract is active.

Here, the ledger entry shares the same TTL as the contract instance. Therefore once the contract instance is no longer active, the data is archived. However, it can also be restored using the `RestoreFootprintOp` operation.

Like persistent storage, instance storage stores data permanently and for the same fees as persistent storage. The key difference is that instance storage allocates a limited storage space.

> Instance storage is best suited for for data that is shared across multiple contract instances but only needs to be accessed during the lifespan of the contract instance. For example, contract admin settings that can be modified by authorized users.

Reviewer: ```suggestion
Instance storage is best suited for data that is shared across multiple contract instances but only needs to be accessed during the lifespan of the contract instance. For example, contract admin settings that can be modified by authorized users.
```

> Instance storage is best suited for for data that is shared across multiple contract instances but only needs to be accessed during the lifespan of the contract instance. For example, contract admin settings that can be modified by authorized users.

Reviewer: > shared across multiple contract instances 

Is this true? How can you share across multiple instances but only access from one instance? I don't think storage is ever shared across contracts... :thinking: this might need some clarity
